### ISSUE TITLE
Bug: Stop-Process does not issue SIGTERM as expected.

### ISSUE THREAD

**Issue by robertbaker** [PowerShell/PowerShell#13664]
## Steps to reproduce

```powershell
taskkill /IM "App With Tray Icon.exe"
```
Without moving mouse, tray icon disappears from tray immediately.

```powershell
Stop-Process "App With Tray Icon"
```
Without moving mouse, tray icon is stays visible. Additionally, starting the process again will display a duplicate icon, the old one disappears when hovering over it.

## Expected behavior

Stop-Process by default should be SIGTERM, graceful.
SIGKILL should only be used after timeout or -FORCE is used.

## Actual behavior
The old tray icon stays visible because app is killed. (This is not actually PowerShell specific, but a technical caveat when processes are killed.)

## Workaround
Use taskkill in place of stop-process for processes that have tray icons.

## Environment data

```none
Name                           Value
----                           -----
PSVersion                      7.0.3
PSEdition                      Core
GitCommitId                    7.0.3
OS                             Microsoft Windows 10.0.20215
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by iSazonov** [comment]
@robertbaker Please check with latest PowerShell 7.1 Preview build.

**Comment by iSazonov** [comment]
I guess it is related to .Net API.
Close as stale issue.

**Comment by PsychoData** [comment]
This seems to be the case that it jumps straight to Kill instead of requesting close

I believe it traces back to here https://github.com/PowerShell/PowerShell/blob/3c7f4b573e96684781b8e1935eeb0b981bcd93bc/src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs#L1250-L1253

or possible here https://github.com/PowerShell/PowerShell/blob/3c7f4b573e96684781b8e1935eeb0b981bcd93bc/src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs#L1399-L1408

It seems like you could send [CloseMainWindow](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.process.closemainwindow?view=netcore-3.1) and since it returns True or False, just fall back to `.Kill()` if `.CloseMainWindow()` if it was not applicable or cannot send and returned False instead. 

This would enable "soft" closes of something like Notepad that has an unsaved document opened in it, as well as -Force closing the same document if desired.

It would also give programs a chance to handle being closed (unless they were Forced) to do things like write save a final recovery copy of unsaved documents, which they won't get if a Kill is sent

**Comment by iSazonov** [comment]
@PsychoData Thanks for your investigations! Do you want to pull PR? I'd review and merge.

**Comment by PsychoData** [comment]
not sure I would be best for it - this would be my first foray into Posh Core coding and in a pretty core area

Not sure if there are any other best practices I should hit for it too, or creating tests for it. 

If someone else wanted to go for it, I would feel better about it

**Comment by iSazonov** [comment]
@PsychoData 
I see you are already reading the code. If you are wasting your time then you have interest. If you have any interest then welcome to contribute!
Some parts of PowerShell are really confusing, but you can contribute to the parts that you are interested in and that you understand. You could read about Working Groups 

The change you propose is not complex. I don't think we can create a reliable test for this - I think manual testing the scenario enough (with good comments in code).

**Comment by PsychoData** [comment]
well - fine if you twist my arm :)

**Comment by PsychoData** [comment]
@iSazonov  there's what I threw in
ideally, I would be thinking some sort of .... timeout.... like `Stop-Process -timeout 500` and it sends a "close" request at first and if it hasn't closed within the timeout, then it sends a Kill, but that is beyond my current C# skills

**Comment by iSazonov** [comment]
@PsychoData Thanks for your contribution!

Original behavior on the cmdlet is silently kill a process. We can not change the default behavior otherwise this will be a breaking change. I mean `Stop-Process notepad.exe` shouldn't wait while ask an user to save a file. So all we can do here is to make the cmdlet more smart and before calling Kill() we could call CloseMainWindow(). If the method blocks current thread we should run it in async, wait some ms and fallback to Kill() if the process is still running. This is a rough description, you could think more and experiment.

In separate PR we could add new parameter like `-GracefullyShutdown [timeout ms]` so that change the default behavior and allow an user interaction and wait until timeout.

**Comment by PsychoData** [comment]
The Powershell process doesn't wait while it asks the user to close, that thread moves right along and leaves Notepad.exe to finish closing itself, or cancel closing itself

You can see the prompt in Powershell is returning immediately and not waiting for the save/noSave/cancel prompts 
Running the second time while the save prompt is up in a Modal window, it can't send the CloseMainWindow again, so it goes through as a Kill on that window, but still send Close to that one Window I had cancelled the previous close

and finally, the Stop-Process -Force closes directly as Kill 
but in any of the cases the pipelines are sending a close or kill command and then moving on
![image](https://i.imgur.com/I6b1SMA.gif)

**Comment by iSazonov** [comment]
On further thought, I see that the current proposed implementation is a breaking change. I thing we should avoid a breaking change and follow the original behavior on the cmdlet that is silently kill a process.
However, I suppose we could make two improvements.
1. Improve base behavior and first try gracefully close process
2. Add new graceful shutdown behavior with new parameter like `-GracefullyShutdown [timeout ms]`
---
For 1:
1.  Call CloseMainWindow() and WaitForExit(20ms) (20ms here is an empiric value.)
2. If the process is still running we call Kill()
---
For 2:
1. If GracefullyShutdown == 0 we call Kill()
2. Otherwise, only call CloseMainWindow() and WaitForExit(<timeout>)

/cc @mklement0 What do you think?

**Comment by PsychoData** [comment]
I am not crazy about the `-GracefullyShutdown` name it makes me think that you have to specify that parameter for it to GracefullyShutdown

Well, if we are going to keep it with force close (Kill) function on `Get-Process Notepad | Stop-Process`  then it would make sense to do something like `Get-Process Notepad | Stop-Process -CloseTimeout <Millis>`  and default to like ..... 200 milliseconds? 

and then it could iterate through the Notepad Processes, start the thread timer for the **Kill** and then send the **Close** signal. 
On Expiration of the "Kill" timer, it can check if it has stopped and kill if needed 

I suppose we could do some testing - but I'm not sure 20 milliseconds would be enough time to the random processes to handle the Close event, close anything down (remember, some of that might be trying to send network session close requests, etc). But that's just deciding on a good default without tying up the timeline for too long. 

For example, sometimes Outlook can close in a fraction of a second, sometimes it takes 1/2 a second, sometimes a few seconds, or sometimes it hangs for what seems like unlimited amount of time (I've seen hours) 

a param like `-CloseTimeout`  might be a good medium to preserve the existing behavior of forcing the process to close, but giving it an (adjustable) timeout to shut itself down before killing. If they specifically need it to **Kill directly** instead of Close **then Kill**,  then I think **-Force** should still skip the **Close** and should go to **Kill** directly still, as it currently does.

**Comment by iSazonov** [comment]
> Well, if we are going to keep it with force close (Kill) function on `Get-Process Notepad | Stop-Process` then it would make sense to do something like `Get-Process Notepad | Stop-Process -CloseTimeout <Millis>` and default to like ..... 200 milliseconds?

Yes, I believe it is mandatory requirement to avoid a breaking change. I like the idea about a default value for the new parameter. As for the parameter name, perhaps we could use `GracefullyStopTimeout` name.

> I think -Force should still skip the Close and should go to Kill directly still, as it currently does.

It would be again a breaking change for Force parameter (today it allows only to close a process of _other_ users). I believe we could use zero timeout (in GracefullyStopTimeout parameter) for the scenario - it is easy understandable for users than complicating Force parameter.

**Comment by mklement0** [comment]
@iSazonov, I agree regarding the concerns about the breaking changes, but let us take a step back:

Conceptually, we are dealing with two modes of termination:

* _Cooperative_, via `SIGTERM`: the process is asked to shut _itself_ down, allowing it to clean up, but it may or may not honor the termination request.

* _Forceful_, via `SIGKILL`: the process is forcefully terminated, without any awareness of that act.

As for default behavior:

* `taskkill` on Windows and `kill` on Unix default to `SIGTERM` - which, given the word "kill" in their name, is unfortunate, but that ship has sailed a long time ago.

* `Stop-Process` defaults to `SIGKILL`

This discrepancy is unfortunate, but I think we're stuck with it.

In terms of terminology, it is similarly unfortunate that "stopping" a process (`SIGSTOP`) in the Unix world means _suspending_ (pausing) it, with the option to resume (continue) it later (`SIGCONT`).


As for synchronous vs. asynchronous behavior:

_Both_ modes of termination are inherently _asynchronous_ - they send a signal and return instantly:

*  _In practice_, the `SIGKILL` method _typically_ acts quasi-_synchronously_, because the _kernel_ handles the termination.
  
    * However, according to the comments at https://stackoverflow.com/q/8762228/45375 re Unix platforms, "the process won't terminate immediately. If a process is doing a system call, it will only end when the system call returns. So if it does some real heavy processing, it may take time.""
  
    * Also, on Unix-like platforms termination may even _fail_ altogether if the target process is in an uninterruptible state. 

* By contrast, `SIGTERM` is much more likely to be _asynchronous_ in practice, since the process _itself_ processes the signal, processing doesn't happen until the process is given CPU time next, and the process may block termination indefinitely with a modal confirmation prompt, or generally refuse to terminate altogether.

----

Given the above, I think the conceptually cleanest approach would be:

* Leave the existing behavior as is.

* Introduce two new switches:

  * `-Request` or `-ByRequest` (as a shorter alternative to `-GracefullyShutdown`; name negotiable), which sends `SIGTERM` instead of the default `SIGKILL`.
  
    * As an aside: It's worth considering support for sending _arbitrary_ signals to processes. On Unix, this is somewhat counterintuitively done via the `kill` utility as well, so you could argue we should extend `Stop-Process` analogously, but  I don't know to what extent that makes sense on Windows, and it is a separate debate.

  * `-Wait`, which waits _indefinitely_ for the process(es) to terminate, irrespective of whether termination is performed via `SIGKILL` or `SIGTERM`.

In case a _timeout_ is needed, use `-PassThru` and pipe to [`Wait-Process -Timeout`](https://docs.microsoft.com/powershell/module/microsoft.powershell.management/wait-process).
This follows the pattern of `Start-Process`, which has a `-Wait` switch, but lacks a way to specify  a timeout.

Another aside: It is unfortunate that the current `-TimeOut` parameters across the cmdlets that support them are `[int]`-typed, allowing for whole-second wait times only. We should consider changing the type to `[double]` instead to support fractional seconds (instead of a  separate `-Milliseconds` parameter, as was implemented for `Start-Sleep`), which I wouldn't expect to be a [Bucket 3: Unlikely Grey Area](https://github.com/PowerShell/PowerShell/blob/master/docs/dev-process/breaking-change-contract.md#bucket-3-unlikely-grey-area) change.

**Comment by PsychoData** [comment]
right, re: the nomenclature, I was usually referring back to the [.NET .Kill()](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.process.kill?view=net-5.0) method which will force-stop immediately and is somewhat analogous to `SIGKILL`. I am not very familiar with Unix or it's Kill method, so I didn't realize the `kill` binary would also send more types of terminations too. 

To keep things clear I'm going to call it `SIGKILL` and `SIGTERM` from here on out, though I could not find a 100% certain source that [.NET .Kill()](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.process.kill?view=net-5.0) equates to `SIGKILL`, or more particularly that [.NET .CloseMainWindow()](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.process.closemainwindow?view=net-5.0) would equate to `SIGTERM` 


On a separate point, it would probably be useful to get some extra information about what kind of terminations it was able to send (`SIGKILL`/`SIGTERM`). 
Example psuedocode in Begin, Process,End mentality for Stop-Process : 
```
Get-Process  Notepad 
<3 Processes found>
Get-Process  Notepad | Stop-Process -CloseRequest 400
Begin: <nothing relevant here?> 
Process: Sent Close request to Process1
Process: Start Timer for CloseRequestProcess1
Process: Sent Close request to Process2
Process: Start Timer for CloseRequestProcess2
Process: Sent Close request to Process3
Process: Start Timer for CloseRequestProcess3
End: Wait for timers to expire OR all processes to be closed, and Kill any Processes that haven't stopped yet
End: Process2 closed itself succesfully
End: `CloseRequestProcess1` and `CloseRequestProcess3` timer expired
End: Send `SIGKILL` to Process3
End: Send `SIGKILL` to Process3
End: Wait for any Timeout to expire
```

I don't know if there is a way to do the "wait and fall back to SIGKILL" `End` section without it having to block the pipeline up for the `SIGTERM` timeout to expire (unless it is `-PassThru`, obviously) but ideally, I would like to let it just keep going asynchronously and not stop up the pipeline. 
If we can have it fall back to `SIGKILL` after a timer/timeout expires without blocking the pipeline, then we could also potentially make the CloseTimeout much larger, say 500 milliseconds, to allow for things like Outlook to close. Otherwise I think any timeout ought to stay much lower to keep blocking the pipeline to a minimum

**Comment by PsychoData** [comment]
On the sidenotes, I think enabling the other types of signals to send besides SIGKILL and SIGTERM sound like a great idea, but most likely seem like it ought to be a whole separate cmdlet?

Converting the `-TimeOut` to double would be nice, or maybe separate `-TimeoutMilliseconds` 

But those should probably be separate issues to talk about those suggestions

**Comment by iSazonov** [comment]
@mklement0 @PsychoData Thanks for sharing your thoughts!

>  I think enabling the other types of signals to send besides SIGKILL and SIGTERM sound like a great idea, but most likely seem like it ought to be a whole separate cmdlet?

.Net does not support signals at all - it is fundamental limitation. We shouldn't go in the direction (until something will be changed in .Net.)

> Converting the -TimeOut to double would be nice, or maybe separate -TimeoutMilliseconds

It would confusing users. We use Timeout parameter of int type in some cmdlets. See `Get-Command -ParameterName Timeout | gcm -Syntax`. _I don't think we should discuss this here._

> -Wait, which waits indefinitely for the process(es) to terminate, irrespective of whether termination is performed via SIGKILL or SIGTERM.

This functionality is in Wait-Process. We have no need to move it to Stop-Process.
----

Main question in the issue is should we try to make the cmdlet more smart on Windows so that call CloseMainWindow() before fallback to Kill()? If no this simplify all. Now I think this would be the best way. Can you vote for this? 
In this case, we would just make two parameter sets- (1) Kill as default, (2) Terminate (or GracefullyStop) for new functionality.

**Comment by mklement0** [comment]
@iSazonov, I agree regarding the issues _not_ to discuss here; it is exactly why I called them _asides_: something to perhaps inspire a separate discussion, though I get that that's problematic without actually creating and/or pointing to such separate discussions, because the temptation is there to respond _here_. To close the one tangent: Point taken re signals in general, but see below.

@PsychoData, re nomenclature: I should have made it clearer that I used `SIGKILL` or `SIGTERM` loosely, as shorthand to refer to the two termination modes I've described.

In terms of implementation, `SIGKILL` (forced termination) corresponds to `.Kill()` in .NET both on Windows and on Unix; `SIGTERM` (cooperative termination) corresponds to `.CloseMainWindow()` in .NET on _Windows_, but on Unix _we'd have to send the actual `SIGTERM` signal_ to get equivalent functionality - and it sounds like we'll have to go native for that, correct?

---

> This functionality is in `Wait-Process`. We have no need to move it to `Stop-Process`.

Yes, we have `Wait-Process`, but just like `Start-Process -Wait` and `Receive-Job -Wait` exist - despite the existence of the dedicated `Wait-Process` and `Wait-Job` cmdlets - implementing `-Wait` on `Stop-Process` would be a _convenience_ switch to make the _typical_ use case easier to implement; more fine-grained waiting - notably via `-TimeOut` - would then require explicit use of `Wait-Process`.

In other words: For _convenience_,  `Stop-Process -ByRequest $somePid -Wait` would be the equivalent of 
`Stop-Process -ByRequest $somePid -PassThru | Wait-Process` - just like it works for `Start-Process`.

---

> Main question in the issue is should we try to make the cmdlet more smart on Windows so that call CloseMainWindow() before fallback to Kill()? If no this simplify all. Now I think this would be the best way.

I see two basic approaches:

**Option A:** Focus on separation of concerns, as suggested by @iSazonov and also in my previous comment:

This means _not_ implementing any fallback logic and _not_ implementing any timeout in `Stop-Process`.

You'd get asynchronous behavior by default, as currently, but can _opt-in_ to wait _indefinitely_ with `-Wait`. Otherwise, use `-PassThru` and pipe to `Wait-Process`

In terms of syntax, this means (for brevity I'm only showing the `-Id`-based input; name of `-ByRequest` negotiable):

```
Stop-Process [-Id] <int[]> [-PassThru] [-Force] [-ByRequest] [-Wait]
```

That is, to implement a timeout with cooperative termination you'd have to use something like:

```
Stop-Process $somePid -ByRequest -PassThru | Wait-Process -Timeout 1
```

`Wait-Process` reports a non-terminating error for (each) target process that doesn't terminate in the timeout period.

This means that in order to fall back to _forced_ termination you'd have to handle that error and call `Stop-Process` again, this time without `-ByRequest`.

The question is how _common_ this scenario is. If it is common, we should make things easier, in which case my suggestion is to add a `-StopOnTimeout` switch to _`Wait-Process`_, which would call `.Kill()` after the timeout has elapsed, followed by `.WaitForExit()` - or go with Option B (see below).

---

**Option B:** Focus on high-level logic, along the lines of @PsychoData's proposal:

```
# FORCED termination
Stop-Process [-Id] <int[]> [-PassThru] [-Force] [-Wait]

# COOPERATIVE termination
Stop-Process [-Id] <int[]> [-PassThru] [-Force] -ByRequest [-Wait] [-Timeout <double>]
```

That is, asynchronous behavior would remain the default (also with `-ByRequest`), unless you specify `-Wait`.

Only with `-ByRequest` (cooperative termination) do you get to specify a timeout, after which `.Kill()` and `.WaitForExit()` are called.

(The assumption is that `.Kill()` either succeeds - in which the process does get terminated and _indefinite_ waiting is guaranteed to end (shortly) - or fails altogether, so that there's no point in supporting a timeout).

---

I can see arguments for both options; if needing to fall back to forced termination is a common scenario, I can see the appeal of option B.

**Comment by PsychoData** [comment]
Well, `.Kill()` returns void, so it either presumably succeeds in sending the signal and ends/returns, or throws an error that it couldn't send the .`Kill()` signal. 

Originally my thinking was a third, I will call Option C

**Option C:** Have Stop-Process send `SIGTERM`/`.CloseMainWindow()` by default, and only send `.Kill()` if it failed to signal "Close" or if `-Force` was specified then default to `.Kill()` 

After all the discussion I see why we wouldn't want to change that functionality of expecting Stop-Process to always result in a process that is `.Kill()`/Forced to close if necessary, since someone may be expecting that for their existing code that was in place - Pester Tests for example. 

my thinking with Option B is that we preserve the high-level function of `Stop-Process` that it will forcibly kill something if necessary, without having to specify -Force for it to `.Kill()` a process if necessary, while only adding a small (minimum) and adjustable timeout. 

Option B is a nice medium between the current "Always default to kill everything" and my [originally provided code](https://github.com/PowerShell/PowerShell/issues/13664#issuecomment-767904909) of "Try to send Close, but Kill if .NET says that failed to send"

**Comment by PsychoData** [comment]
I think the only point we are getting hung on talking about is that I am thinking 
`Stop-Process 'Notepad','Outlook'` (with a default -ByRequest implied)
Would result in 
 * Send `Outlook` and `Notepad` a **Close** signal (repeat for any additional processes specified)
     * Start ByRequest timer here
     * This is starting all of the ByRequest timers as close as possible in time
     * Still attempt to send `.Close()` but if that fails jump straight to `.Kill()` 
 * Wait until `ByRequest` timeout expires in the End section of the processing
     * Loop until their ByRequest Timers have expired, or the processes are all stopped
     * Could Optionally add a "NoWait" or just turn ByRequest timeout to 0 or 1 milliseconds 
     * If timeouts expire, and there are still valid Processes, send the `.Kill()` and let the method return/continue onward

In many... probably most... cases this timeout would not even be needed because the program will handle being sent a `.Close()` signal and just close itself before you're checking on the timeout anyway. If this used a Low default  for ByRequest (20 millis? 50? 100? No idea of a best value for that yet) then we should be looking at. 

Keep in mind, that if we are running on 5 processes with a 500 millis ByRequest timeout, this won't be 500+500+500+500 it would be around 500 + time to send all the `.Close()` signals. For me, using three processes, that took about 100 millis to send the `.Close()` commands and finish the cmdlet
![image](https://user-images.githubusercontent.com/3719116/106511648-82fa8a80-649e-11eb-99eb-118f7ad3868a.png)

So if it had a 500 milli ByRequest timer, you are done sending the `Close()` signals long before the ByRequest time expires... and I would figure it took maybe .... 600 millis total execution time with the `send close()`, `Wait for ByRequest to expire`, and `Some Processes did not gracefully close, Kill them`  depending on how busy the machine was and could revisit the timer thread

**Comment by iSazonov** [comment]
> Yes, we have `Wait-Process`, but just like `Start-Process -Wait` and `Receive-Job -Wait` exist - despite the existence of the dedicated `Wait-Process` and `Wait-Job` cmdlets - implementing `-Wait` on `Stop-Process` would be a _convenience_ switch to make the _typical_ use case easier to implement; more fine-grained waiting - notably via `-TimeOut` - would then require explicit use of `Wait-Process`.

In the case this could be separate enhancement since it is not mandatory for enhancement we consider here.

> I should have made it clearer that I used SIGKILL or SIGTERM loosely, as shorthand to refer to the two termination modes I've described.

I feel most of users follow intuitively the terms. I believe we need to follow this in parameter names too. If we ask users what is:
```powershell
Stop-Process -Kill
Stop-Process -Terminate
```
most of them give us right description. But `Stop-Process -ByRequest` will force them to read docs before get understanding its semantic. 

If we start with adding new Terminate switch (and perhaps Kill for symmetric) this will address current issue in simplest way and open ways for future enhancements.

**Comment by mklement0** [comment]
@iSazonov:

On a meta note, I think at this point it is clear that before implementing anything we need to write up a _new_, focused proposal, following this discussion.
Therefore I think it's worth having this discussion, even as it gets lengthy.

---

Re `-Wait` switch:
> In the case this could be separate enhancement since it is not mandatory for enhancement we consider here.

Based on my new proposal below we won't need the `-Wait` switch, after all, but _in general_ I find that argument problematic: While with Option A, `-Wait` would _technically_ not be necessary (it would be with Option B),  it would greatly increase the utility of the enhancement. If `-Wait` were still in the picture, I wouldn't see a reason to do this separately, not least because I suspect it would then never be implemented.

---

> `Stop-Process -Kill`
> `Stop-Process -Terminate`

While I like the idea of these contrasting switches to make the two modes explicit, there is the awkwardness of then having a switch being _true by default_, namely`-Kill`.

Also:

* Familiarity with the term _terminate_ likely only applies to _Unix_ users; in the context of the GUI window-centric Windows, _close_ is used, as reflected in the `.CloseMainWindow()` method name; also, _terminate_ by itself, from a natural-language perspective, doesn't adequately convey the aspect of cooperative / graceful stopping.

* (_Kill_, while unambiguous in terms of common usage, is tainted by the conceptual confusion of both `taskkill.exe` and `kill` exhibiting _terminate_ / _close_ rather than kill behavior by default, but there's obviously nothing we can do about this.)

* _Stop_ invariably means something different in PowerShell, which in signal terms in Unix (`SIGSTOP`) - regrettably - means _suspending_ a process.

  * As an approved verb, `Stop` has no precise definition with respect to _cooperative_ vs. _forced_, but in practice `Stop-Process` is the only outlier in that it performs _forced_ stopping (`Stop-Job`, `Stop-Computer`, `Stop-Transcript` do not).
  
  * However, given that the [approved-verbs documentation](https://docs.microsoft.com/en-us/powershell/scripting/developer/cmdlet/approved-verbs-for-windows-powershell-commands) lists under "synonyms to avoid" for `Stop` "End, Kill, Terminate, Cancel", the implication is that `Stop` is meant to cover _both_ cooperative and forced stopping.

  * Of course, it would then make sense to _consistently_ make cooperative stopping the default, with an _opt-int_ for forced stopping. Regrettably, `Stop-Process` not only doesn't offer cooperative stopping, but _invariably_ performs forced stopping.

In short: We won't be able to use existing terminology from one platform without it clashing with that of another.

My suggestion was motivated by using names based on platform-neutral _abstractions_ that express the _conceptual intent_; perhaps `-ByRequest` isn't great, and something like `-Graceful` (a shorter version of the previously suggested `-GracefullyShutdown`) is more descriptive.

_If_ we had established semantics of `Stop` _without further qualification_ implying _cooperative_ (graceful) stopping, then we'd only ever need a switch to opt-into _forced_ stopping - and the obvious name for such a switch would be `-Force`.

---

@PsychoData:

> (with a default `-ByRequest` implied)

We cannot default to cooperative stopping without breaking backward compatibility - users may have come to rely on _unconditional_, forced, quasi-synchronous termination.

Given the conceptual musings above, I sincerely wish we _could_ break backward compatibility, which makes this a candidate for #6745.

As for timeouts: For simplicity and predictability, I'd use the timeout as a single, overall waiting period, irrespective of how many processes are targeted: I would start a _single_ timing _in the `End` block_ (not sure what `Wait-Process` does).

---

Let me propose **Option C**:

* Make `Stop-Process` _synchronous_ by default, with an _opt-in_ kill timeout that _only applies if `-Graceful` (`-ByRequest`) is also specified. This is similar to Option B, except that `-Wait` is no longer required.

* Conversely, a new `-NoWait` switch be used to request _asynchronous_ behavior.

Note: For consistency, I suggest also making the by-default _kill_ operation synchronous (call `.WaitForExit()` after `.Kill()`): this would technically also be a breaking change, albeit an _acceptable_ one ([bucket 3: Unlikely Grey Area](https://github.com/PowerShell/PowerShell/blob/master/docs/dev-process/breaking-change-contract.md#bucket-3-unlikely-grey-area)): I think it would have virtually no impact on existing code (given that `.Kill()` forcefully terminates, I would expect the extra time spent waiting for actual termination to be negligible) while making the behavior slightly more predictable (even though it's unlikely that the current asynchronous behavior surfaces as such).

```
# FORCED termination, (now) synchronous by default, except if -NoWait is passed.
Stop-Process [-Id] <int[]> [-PassThru] [-Force] [-NoWait]

# COOPERATIVE (graceful) termination: synchronous (indefinite wait) by default, except if -NoWait is passed.
Stop-Process [-Id] <int[]> [-PassThru] [-Force] -Graceful [-NoWait]

# COOPERATIVE (graceful) termination: synchronous, but with optional timeout resulting in
# *synchronous killing*  if the processes don't terminate in time.
Stop-Process [-Id] <int[]> [-PassThru] [-Force] -Graceful [-Timeout <double>]
```

@PsychoData, note that this proposal intentionally does _not_ include a built-in, automatic kill timeout with `-Graceful` termination: Users should be free to decide whether they want to:

* wait indefinitely (by default), which is the safest behavior.
* not wait at all (`-NoWait`)
* wait for a while and, in case of non-termination within that period,  _kill_ (synchronously) (`-TimeOut <n>`) 
* wait for a while and, in case of non-termination, only report _an error_, via `Wait-Process`: 
`Stop-Process $somePid -Graceful -NoWait -PassThru | Wait-Process -Timeout $n`

**Comment by iSazonov** [comment]
@mklement0 for information - Terminate() method exists in WMI Win32_Process class, I guess it _kills_ a process. :-)

**Comment by iSazonov** [comment]
Based on @mklement0 comments I suggest follow final proposal:

Enhance the cmdlet with -Graceful / -Gracefully switch which will call CloseMainWindow() method (and fallback to Kill() if the method returns fasle). This will gracefully terminate the process on Windows. No changes of a functionality on Unix.

All other enhancements we have discussed here we will consider later.

**Comment by mklement0** [comment]
Thanks, @iSazonov, but let me spell out the implications of your proposal, from which I conclude that it is _not_ worth implementing as such:

* The behavior will be inconsistent and potentially ineffective, because basing the behavior on whether `CloseMainWindow()` returns `$false` is insufficient - see below.

* The solution will only work on Windows - even though sending `SIGTERM` would amount to equivalent functionality on Unix.

---

Note: I am partly out of my depth here, but I hope I'm at least fundamentally correct:

* `CloseMainWindow()` returns `$false` only in two cases:

   * The application _does_ have a (main) message loop (it is a GUI-subsystem application) but is _currently_  in a state where it cannot process messages - a typical example is a modal dialog currently being shown.
   
   * The application does _not_ have a message loop (a console-subsystem application), except if it has one _indirectly_, by running _directly_ in a _console window_, which itself _does_ have a message loop. This means:

     * A console application running _directly_ in a console window - typically a _shell_ - will typically respond with `$true` _via its console window's message loop_, and it is the console window that then terminates the application.

     *  By contrast, a console application launched _from a shell_ has _no_ associated window and always returns `$false`.

       * In either case, a console application can optionally register for console events such as `CTRL_CLOSE_EVENT`, via the [`SetConsoleCtrlHandler`](https://docs.microsoft.com/en-us/windows/console/setconsolectrlhandler) WinAPI function, which allows them to stop gracefully, analogous to a process opting to handle `SIGTERM` on Unix. 
 
---

This implies for your proposal that if `-Graceful` is specified:

* All console applications (except shells) will effectively always be _killed_ instead (as they always return `$false`).

   * To fix that, the target process' controlling console window would have to be identified and the `.CloseMainWindow()` method must be called on _it_ - only that would give such console applications a chance to terminate gracefully, via their handlers.

* GUI applications may end up _not terminated at all_, because returning `$true` only indicates that the message was _processed_, not that termination will actually be performed.

   * In fact, a GUI application whose message loop is responsive may put up a modal dialog _in response_ to `.CloseMainWindow()` (as in the Notepad example above) and indefinitely wait for user input. From PowerShell's perspective, such a `Stop-Process -Graceful` call would then amount to _quiet failure_: the process isn't terminated, and no error is reported.

---

Even if we address all the problems above - i.e. if we truly give all all targeted processes a chance to terminate gracefully - _enforcing_ (ultimate) termination should (a) be opt-in and (b) can, as stated, only be achieved by _waiting for actual termination based on a timeout_, given that `SIGTERM` / `CTRL_CLOSE_EVENT` handlers can _refuse_ to terminate in response to a request.

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee reviewed this and agrees to not make a breaking change where automation will expect processes to be killed.  .NET currently does not provide a way to send `SIGTERM` on Unix systems and no corresponding capability on Windows (where `CloseMainWindow()` is similar, but not the same as noted for console processes).  This may be better served by a community module/cmdlet that is OS specific.

**Comment by iSazonov** [comment]
> Even if we address all the problems above - i.e. if we truly give all all targeted processes a chance to terminate gracefully - _enforcing_ (ultimate) termination should (a) be opt-in and (b) can, as stated, only be achieved by _waiting for actual termination based on a timeout_, given that `SIGTERM` / `CTRL_CLOSE_EVENT` handlers can _refuse_ to terminate in response to a request.

I believe we should do the cmdlet too smart and complex. 
It is currently impossible to send an event or signal to a process so that it terminates gracefully.

_The suggestion is to just add such a feature - just send the signal and nothing else._

All other smart things the user can do himself (or we can add later after receiving feedback).

---
We can implement this on Unix too https://stackoverflow.com/questions/41041730/net-core-app-how-to-send-sigterm-to-child-processes

**Comment by mklement0** [comment]
Thanks, @iSazonov - good find, and I do think that starting small is an option, but let me flesh your suggestion out to see its full implications:

* Existing `Stop-Process` behavior will remain as-is:  unconditional _kill_ behavior, _asynchronous_ (even though in practice it will _in effect_ typically be synchronous).

  * Unless killing _fails_ (such as due to lack of permissions), it can safely be assumed that the target processes _will_ terminate (and may (likely) or may not already have, by the time the next PowerShell statement executes).

* Introduce only _one_ new switch, `-Graceful`, which does the following:

  * On Unix, it sends signal `SIGTERM`, and on Windows it calls `CloseMainWindow()`.
  * If sending the signal fails (typically due to a permissions problem) or `CloseMainWindow()` returns `$false` (see below), report a _non-terminating_ error (per process).

Either way (after either failing or succeeding to send `SIGTERM` /  call `CloseMainWindow()`), `Stop-Process`'s job is done, which means:

* Termination may not occur at all (only in the _failure_ case do you have _instant certainty_), because the target processes may ignore the signal / close request.

* If termination does occur, it may not occur until some time later (more noticeably so than with the default _kill_ behavior).

To detect whether termination occurred, a separate `Wait-Process` call is needed, sensibly with a `-Timeout` argument.
If the processes don't terminate within the timeout period _and termination must be enforced_, another `Stop-Process` call is needed, this time _without_ `-Graceful`, to effect killing.

---

**When use of `-Graceful`  is then meaningfully supported**:

* On **Unix**-like platforms:
  *  **_Any_ process** can be targeted; if a given process doesn't have a `SIGTERM` handler, the system will kill it.

* On **Windows**:
  * **GUI applications** can be targeted.
  * (Console-based) **shells** can be targeted; more accurately, any console application that _directly_ has a console (window) attached to it.

This means that **console applications launched _from a shell_ can NOT be targeted** (such as `node.exe` started from a shell with a script that runs a local webserver, for instance):

* `CloseMainWindow()` invariably returns `$false` for them, so `Stop-Process` will report a non-terminating error.
  * Unless we want to go to the trouble of detecting the application type (GUI vs. console), the error message should mention both potential failure reasons: the target is either (a) a GUI app that is in a state where it cannot process messages or (b) an app that has no associated window and therefore no message loop, such as a console app (with no console directly attached).

*  It seems that Windows currently fundamentally doesn't support sending a terminating request (`CTRL_C_EVENT`, `CTRL_BREAK_EVENT`, via [`GenerateConsoleCtrlEvent`](https://docs.microsoft.com/en-us/windows/console/generateconsolectrlevent)) to a _single_ process running (indirectly) inside a console - except if it was explicitly created as part of a process _group_; see https://github.com/microsoft/Terminal/issues/335 for (very in-depth) background information.

   * The only way to get such processes to (potentially) terminate gracefully is if the (indirectly attached) _console window as a whole_ is closed, but that is inappropriate, if only a single process running inside that window is to be targeted.

---

If everyone agrees that this - initially minimal - functionality is still beneficial and the implications are understood and well-documented, I think it's worth doing.

**Comment by iSazonov** [comment]
@mklement0 I think it makes no sense for us to try to do something too clever since not even the platforms themselves do it. (Moreover, there are differences in async/sync behavior.)
If SIGTERM always fallback to SIGKILL then we are forced to follow this. The same should be the case on Windows - if `CloseMainWindow()` returns false then just call `Kill()`.

```c#
if (Graceful.Present && TryStopProcessGacefully())
{
    // return;
}
else
{
    process.Kill();
}
...

void TryStopProcessGacefully()
{
#if UNIX
    SendTerminateSignal(process);
    return true;
#else
    return CloseMainWindow();
}
```

For reference - Process.Kill() on Unix to implement SendTerminateSignal() https://source.dot.net/#System.Diagnostics.Process/System/Diagnostics/Process.Unix.cs,57

**Comment by mklement0** [comment]
> it makes no sense for us to try to do something too clever

I was proposing the very opposite (which is what I thought you were advocating too): 
I was proposing to simply use the underlying platform feature, without trying to superimpose any additional logic - and it looks like your pseudo code is attempting just that, so _I don't think there is any disagreement here_.

A few quibbles and clarifications:

> `SIGTERM` always fallback to `SIGKILL`

* `SIGTERM` doesn't _fall back_ to termination by the system:
   * Termination by the system is the _default behavior_ that any process may _choose to modify_ by implementing a signal handler.

* Similarly, `.CloseMainWindow()` doesn't _fall back_ to closing the window:
  * The system closing the window is the _default behavior_ that any process (that actually has a main window) may _choose to modify_ by responding to the close message.
  * Calling `.CloseMainWindow()` on a process for which _there is fundamentally no window to close_ - i.e., one without a message loop (such as a console app launched from a shell) - predictably and sensibly fails (`$false`).

In the _success_ case - being able to _send_ the signal / close message - the uncertainty over whether that signal / message will _eventually, possibly asynchronously_ be  _honored_ is built into both mechanisms.

_Not_ trying to resolve this uncertainty through superimposed logic is the gist of the [previous proposal](https://github.com/PowerShell/PowerShell/issues/13664#issuecomment-774185156) and seemingly yours too.

The only challenge I see is to make users understand the limitations of what processes can be targeted on Windows.

**Comment by mklement0** [comment]
@iSazonov: Sorry, I misread your previous comment: there _is_ a fundamental disagreement here: I think we should _not_ fall back to killing, for the reasons stated.

**Comment by PsychoData** [comment]
The only reason that we started talking about sending SIGterm, but then maybe waiting for a time out of some sort and then sending sigkill was that way we could preserve the high-level functionality that the processes **will be stopped** once the cmdlet is done running. Or if there was some error, throw an exception/error.

There should probably be some option to just send the close event without sending the kill, but foremost we should preserve the existing functionality so we don't break current deployments. 

I really don't think that using a separate wait-process or synchronously waiting with .WaitForExit would be a good idea at all, because that process doesn't have a way to short circuit out if it is taking 20 minutes to close. 


I looked at the code in the dotnet core clr , and there is definitely going to be no benefit to this on Unix currently, but windows still could benefit. 
On the windows side the process for sending the actual commands to the processes currently only lists like three commands - close, kill, stop, resume - And it looks like that might could be extended, but that would be a separate argument for a separate repo with a separate team most likely. 

I'll try to hack together some code to demo the functionality, because I feel like all of these other extensions that have been discussed could certainly be useful, but the feature bloat from the original goal is significant.

If the dotnetCLR finally gets updated to have support for sending more types of events (preferably including arbitrarily sending whichever signal we want - like the Unix kill command) then that seems like it would be the time to revisit this and abstract the sending signals to some other function possibly

**Comment by mklement0** [comment]
> that the process _will be stopped_

* With _killing_ (the current behavior), you get that for free with the current behavior (although not strictly synchronously), but _not by explicit high-level logic design_, but by a straight pass-through to the underlying system functionality.
 
* With (potentially cooperative) termination,  the underlying system functionality offers _no_ stopping guarantee.

> There should probably be some option to just send the close event without sending the kill

That's the current proposal: `-Graceful` by itself would just be a platform-abstracted way to send `SIGTERM` on Unix, and to call `.CloseMainWindow()` on Windows.

I definitely would like to see the high-level functionality of _ensuring_ that the process is stopped, but the above would be a fairly simple and straightforward start _whose behavior doesn't deviate from the underlying system mechanisms_.

If `SIGTERM` / `.CloseMainRequest()` _defer_ to the target process with respect to whether it actually terminates, so should we _by default_.

---

As for  a (possibly later) enhancement that _builds_ on the above:

First, I agree that general signal support should be a separate discussion.

I'm always a fan of _desired-state_ functionality, but I believe it should be _opt-in_ here, to modify the underlying system behavior _on demand_.

If you're _asking_ a process to terminate, it is _not_ a given that your intent is to _kill_ it, if it refuses to / doesn't terminate within a given timeout.

> I really don't think that using a separate `wait-process` or synchronously waiting with `.WaitForExit` would be a good idea at all

Of course, using a timeout makes sense to prevent infinite waiting.

But users should have a choice as to:

* what happens on timeout: give up, or kill?

* how long to wait - different timeouts may be appropriate for different processes / scenarios.

If we do go with a _default_ timeout (and I have no idea what period would make sense, but it definitely must be clearly documented), it should only apply if you've _opted into_ fallback-to-killing behavior, with a switch.

That leads me to **Options D and E**:

What they share:

* Retain existing behavior, and make the new `-Graceful` switch asynchronous too: if `-Graceful` succeeds, it just mean that the _request_ was successfully sent, and termination may or may not occur, and there's no guarantee when. (So far, this is the minimal enhancement suggested).
  * `Stop-Process` alone would be PowerShell's cross-platform way of _killing_ a process (which is an unfortunate default, but we're stuck with).
  * `Stop-Process -Graceful` would be PowerShell's cross-platform way of _requesting_ termination - with no guarantee that the request will be honored - and I'm now again tempted to bring the word _request_ in.

**Option D**: _with_ a default timeout for the fall-back-to-killing opt-in:

```
Stop-Process ... -Graceful [-KillOnTimeout [-Timeout <double>]]
```

That is, `-KillOnTimeout` is the opt-in to the kill fallback, and by itself uses a _default_ timeout, which can overridden with `-Timeout` (which must be a nonzero, positive value).


**Option E**: _without_ a default timeout:

This forces users who want to ensure ultimate termination after request-based termination fails to specify a timeout _explicitly_.

```
Stop-Process ... -Graceful [-KillTimeout <double>]
```

**Comment by mklement0** [comment]
Actually I missed that the current behavior actually _does_ make an attempt to gracefully terminate _some_ processes, namely (by definition Windows-only) _services_:

https://github.com/PowerShell/PowerShell/blob/6ad5f65e18e921570711ddea75d4cbd2297de134/src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs#L1245-L1248

(As an aside: the comment describing the `-Force` switch doesn't match its behavior:
https://github.com/PowerShell/PowerShell/blob/6ad5f65e18e921570711ddea75d4cbd2297de134/src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs#L1157-L1164
)

That said:

* On _Unix_, this doesn't apply, and calling `Stop-Process` _always_ means: _send a kill request_.

* On Windows, this accommodation for services _only_ makes for an awkward inconsistency (especially given that there's a dedicated, graceful-shutdown `Stop-Service` cmdlet - but, again, one we're stuck with.

**Comment by PsychoData** [comment]
the goal isn't full-synchronous killing. 

It's just as close as the libraries let us get to async, short of them being rewritten to send the signals directly and just move on

similar to Unix, while maintaining the vanilla `Stop-Process Notepad` capability would be to just let it 
 - send the CloseMainWindow() event - which just sends the message to the Window Handle through Interop services
 - give the process at least a few milliseconds to handle that close request and work on closing (the timeout) 
 - send the Kill event, if our timeout expires and the process hasn't stopped
 - At this point `Stop-Process` is _done_ with it's functionality, and it exits, with the only final action it does currently being to kill the current Process if it was instructed to do so

https://github.com/PowerShell/PowerShell/blob/master/src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs#L1281-L1290

A few Example flows, 

`Stop-Process Notepad` where Notepad has NO UNSAVED text entered and will NOT INTERRUPT the CloseMainWindow() 
 - Processing Starts
 - CloseMainWindow() is sent
 - Timer for Timeout is started
 - Processing ends
 - EndProcessing Starts
 - Loop until (Timer ends or Notepad finishes closing)
 - Notepad will likely finish Closing and the loop will exit
 - Stop-Process ends

`Stop-Process Notepad` where Notepad HAS UNSAVED text entered and WILL INTERRUPT the CloseMainWindow() to prompt the user to save
 - Processing Starts
 - CloseMainWindow() is sent
 - Timer for Timeout is started
 - Processing ends
 - EndProcessing Starts
 - Loop until (Timer ends or Notepad finishes closing)
 - Notepad will NOT stop before the timeout and a .Kill() will be sent to finish the job
 - Stop-Process ends without checking if the process finished exiting

`Stop-Process Outlook`
 - Processing Starts
 - CloseMainWindow() is sent
 - Timer for Timeout is started
 - Processing ends
 - EndProcessing Starts
 - Loop until (Timer ends or Outlook finishes closing)
 - Outlook may finish closing, or timer might expire
 - Stop-Process ends


I don't think this would be the right place to go adjusting the flow of the Stop-Process way of handling the Services stopping. That's a different discussion in a different issue.

**Comment by mklement0** [comment]
> short of them being rewritten to send the signals directly and just move on

The existing mechanisms are:

* asynchronous from an waiting-for-actual termination perspective
* synchronous with respect to waiting until the signal has been sent / the message has been posted, both of which are unrelated to the actual point in time of termination, if any.

> the goal isn't full-synchronous killing.

Once you superimpose (request-termination-then-)fallback-to-kill logic, you get _quasi_-synchronicity even if you don't call `.WaitForExit()` after `.Kill()`, given that killing is near-synchronous.

And, of course, if the process terminates _within_ the timeout period, you have synchronous behavior by definition.

To put it differently: what you're invariably looking for is synchronicity with respect to knowing that actual termination _will occur_ (was successfully _initiated_), which, if a termination _request_ is first sent invariably involves _waiting_.

> I don't think this would be the right place to go adjusting the flow of the Stop-Process way of handling the Services stopping.

Agreed - that's why I said we're stuck with the behavior.

Also, just to remind us, the committee has already turned down any enhancement here, so this may never happen or at least not anytime soon.

I'd say the only chance for this to be revisited is if we agree on a way forward that also addresses the committee's concerns, and present that in a _new_, focused feature-request issue.

The committee's concerns were:

* The change should not be breaking.

  * I'm not sure your proposal qualifies (see below).

* >  .NET currently does not provide a way to send `SIGTERM` on Unix systems and no corresponding capability

  * @iSazonov has demonstrated that it would be fairly easy to implement it ourselves.

* > no corresponding capability on Windows [...] for console processes

  * With a kill fallback this would be _somewhat_ mitigated, but _suboptimally_ in that such (no-directly-attached-console) console applications would then _always_ be _killed_.

  * However, given that Windows itself offers no such capability, you could argue that this best-effort approach is still preferable (and once Windows does offer the ability, the cmdlet could be amended).

---

Do I understand  correctly that you want to bake the request-first-then-fallback-to-kill logic  into `Stop-Process` _by default_?

Even though I can see the appeal of this from the perspective of  _ensuring_ termination, it does constitute a breaking change:

* existing code may rely on _forceful_ termination, i.e. explicitly _not_ giving the process a chance to clean up.

* the invariably necessary waiting period (except, on Windows, if `.CloseMainWindow()` returns `$false`) slows down each call.

Even if everyone were comfortable with this change, you would then need a switch such as `-NoKillFallback` to allow request-only functionality, and there should still be a user-specifiable `-Timeout`.

**Comment by PsychoData** [comment]
I don't know - I just tried to kick something that seemed like a good idea along. 

From [the very beginning](https://github.com/PowerShell/PowerShell/issues/13664#issuecomment-767072983) I was saying I didn't think I was best for this because I knew it would be a breaking-ish change and there would likely need to be considerations for it to preserve the `.Kill()` effect directly by default, even though they had perfectly valid alternatives or fairly minor mitigations, like a single delay of maybe 500 millis for a  `-ByRequest TimeOut`, would probably need additional parameters added (Which I'm not sure how to add in .cs) to bypass the falling back to Kill() after timeout expired, possibly even more. 

------------

But it would be very useful for any program that properly handled a Close request, and give IT a MUCH easier way to gently close processes, without them. 

For example, if I wanted to issue a restart - nothing would stop me from saying `Get-Process | Stop-Process -RequestCloseOnly` or maybe `Get-Process | where {From Logged in User} | Foreach-Parallel { $_ | Stop-Process -RequestCloseOnly }` to close every open process, as long as it wasn't something like Notepad with unsaved text, before sending a restart command. 

-----------

Someone else can try to chase this down if they want, but my effort to get this done is though, because my skill level was spent before I ever made the PR when @iSazonov was [pushing me to](https://github.com/PowerShell/PowerShell/issues/13664#issuecomment-767306906), exactly like I said it would be.

**Comment by iSazonov** [comment]
> > `SIGTERM` always fallback to `SIGKILL`
> 
> * `SIGTERM` doesn't _fall back_ to termination by the system:

I mean if a process doesn't implement SIGTERM handler a stopping behavior will be like SIGKILL. 
Thus, both on Windows (CloseMainWindow() returns a false) and on Unix (no SIGTERM handler) there may be processes that cannot "gracefully stop" and will just kill.
And my code snippet above reflects this.

**Comment by mklement0** [comment]
You're right, but that means that on Unix the `else
{
    process.Kill();
}` branch will never be reached - yet the process may _not_ terminate, if the process _does_ have a `SIGTERM` handler but decides not to terminate.

On Windows _it shouldn't be reached_, at least not _by default_, because that would mean superimposing destructive logic on the underlying system behavior.

With an _opt-in_ such as `-KillOnTimeout` it would be fine, however: `.CloseMainWindow()` returning `$false` could then rightfully be treated like an instant timeout;  however, on _successfully_ sending `SIGTERM` / calling `.CloseMainWindow()`  _you'd still have to wait_, because you otherwise cannot _ensure_ that termination will occur.

If the intent is for `-Graceful` to _default_ to `-KillOnTimeout`, we would need a switch that _opts out_ of the kill fallback - unless you think that that's not needed.
Either way, this would mean a departure from the behavior of the underlying system call, whereas my thinking was that _by default_ we should simply expose the underlying system behavior as-is.

To summarize:

* I think touching the default behavior - unconditional killing - is too much of a breaking change (as unfortunate as that default is).

* This means that termination by request requires at least one opt-in switch, such as `-Graceful`.

 * This leaves the question of whether this switch _alone_ should effect try-gracefully-with-fallback-to-kill logic, or whether it should simply defer to the underlying system calls - which have _no_ fallback.

    * Deferring to the underlying system calls would make `Stop-Process -Graceful` behave the same as `taskkill` on Windows (without `/f`, the kill switch) and `kill` on Unix (without `-9`, the kill switch); in other words: we'd have consistency, albeit with the defaults reversed (unavoidably, for backward compatibility).
     * This means: 
       *  Like `taskkill` and `kill` by default, `-Graceful` (alone) would be happy to _send_ the termination request, and would consider that alone success, without _ensuring_ ultimate termination (that is up to the target process, _if_ it has a termination-request handler).
       *  If sending _fails_, an error would occur, which on Windows would predictably occur for any console application without a directly attached console (any process without a window message loop); `taskkill` provides a meaningful error message in this case: `ERROR: The process "<pname>" with PID <pid> could not be terminated.
Reason: This process can only be terminated forcefully (with /F option).`

**Comment by mklement0** [comment]
@PsychoData 

>  I just tried to kick something that seemed like a good idea along.

Agreed - I do think we should provide this functionality, but the tricky part is _how_.

Thanks for the discussion; even if no immediate action follows, I think it was useful to get clarity.

> Which I'm not sure how to add in .cs

Note that it's perfectly fine to only contribute _conceptually_ to a discussion, without being the implementer or needing to know all technical details.

I hope that it's clear that the sticking point here is the up-front conceptual work - agreeing on the end-user experience and assessing backward-compatibility concerns, and that just happened to turn out much more complex than originally anticipated.

**Comment by iSazonov** [comment]
@mklement0 I think we all are in consensus that we don't change the default behavior of the cmdlet and we all find the new Graceful option being useful.
Perhaps we need to rename Graceful switch to TryGraceful to explicitly say users that the cmdlet does not guarantee that a process will necessarily stop (due to the way the operating systems work). I suppose we don't need to worry about how processes will react to SIGTERM and CloseMainWindow(). This specific can be described in the documentation. I guess my code snippet is the best compromise. I'm sure this will cover most custom scenarios. More complex cases can always be resolved with additional commands.

**Comment by mklement0** [comment]
@iSazonov 

>  think we all are in consensus that we don't change the default behavior of the cmdlet and we all find the new Graceful option being useful.

👍

> Perhaps we need to rename `-Graceful` switch to `-TryGraceful`

👍

> I suppose we don't need to worry about how processes will react to SIGTERM and CloseMainWindow().

That is definitely an option - we can keep desired-state logic out of the cmdlet for now, and possibly enhance later.

> I guess my code snippet is the best compromise.

No, I don't think so, because _by default_ there should be _no_ fallback to killing - just like `taskkill` without `/f` (the equivalent of `-TryGraceful`) does _not_ kill if `.CloseMainWindow()` returns `$false` - instead, it reports an error that _tells you how to KILL if you really want to_.

In terms of your snippet, this means:

```csharp
if (TryGraceful.Present)
{
   if (! TryStopProcessGacefully()) 
  {
     // Report non-terminating error along the lines of:
     // "Graceful termination not possible (the process either doesn't support it at all (no message loop) or cannot process messages in its current state); to kill the process, call without -TryGraceful"
  }
}
else
{
    process.Kill();
}
```

Again, desired-state logic is desirable, but falling back to killing only if the close-request cannot even be _sent_ makes for _half_ of an ensured-termination feature: a successfully sent request may still result in _non-termination_, which means that you haven't _ensured_ termination overall.

A proper ensured-termination feature would require _waiting for termination_ (if the request was _successfully_ sent, otherwise you can kill instantly), which introduces the need for a timeout (at least a default one, but ideally also a user-specifiable one).

(Note that on Unix the case where `SIGTERM` cannot even be _sent_ likely means a fundamental problem, such as insufficient permissions, which means that `SIGKILL` wouldn't work either; therefore, this case should just result in an _error_; this could also happen with `.Kill()` on Windows)

In terms of syntax, this means:

* If we initially go _without_ ensured-termination in the picture we would only need `-TryGraceful`

```
Stop-Process ... [-TryGraceful]
```

* If and when we decide to implement ensured-termination logic: 

```
Stop-Process ... -TryGraceful [-Force [-Timeout <double>]] 
```

* `-Force` is the best choice to signal the intent to _fallback to killing_, and I think it's defensible to give the existing `-Force` switch this semantics _when combined with `-TryGraceful`_.

* Unless `-Timeout` is also specified, a well-documented _default_ timeout would apply before killing is resorted to; this default timeout should probably apply _per input process_, whereas a user-specified one should apply overall.

**Comment by iSazonov** [comment]
@mklement0 If on Windows we can detect whether a process can handle a close event (CloseMainWindow() returns false), on Unix we cannot detect this for SIGTERM. (If we haven't permissions we will get an exception in any case.) I'd prefer to have unified behavior for all OS-s. It is first argument to do not throw and just send a close event.
Second argument is there are not benefits for users to get an error.  If users initiate a grace stopping but don't care whether the process was really stopped the error makes no sense. If users care about that the process was really stopped they should explicitly check this (with Wait-Process) in any case (is there an error or not).

**Comment by mklement0** [comment]
> It is first argument to do not throw and just send a close event.

Sending a close event: yes. 

If the system tells you that it cannot, you report a non-terminating error (rather than _throwing_), just as you would with a permissions problem with `SIGTERM` on Unix).

This amounts to unified behavior _with respect to the underlying system capabilities_.

> Second argument is there are not benefits for users to get an error. If users initiate a grace stopping but don't care whether the process was really stopped the error makes no sense

They may care about being able to _request_ graceful stopping, but leaving it up to the target process to comply (see below).
That is what the underlying system mechanisms provide - even if that is unsatisfying from a _desired state_ viewpoint.

On Unix, the processes _almost always_ comply - from what I can tell, even GUI applications such as `gedit` and `Firefox`, which - unlike GUI applications on _Windows_ - do _not_ pop up a modal dialog if their data is "dirty", and quietly terminate (potentially due to not even having a signal handler).

That is, on Unix, where only truly exceptional conditions (such as lack of permissions) prevent _sending_ the signal, the sending the signal will _typically_ result in termination - but that termination is _not guaranteed_, given that if the process has a signal handler, it may refuse (or the signal handler may crash).

On Windows, you're much more likely to run into non-termination:

* (a) applications without a message loop (such as console applications called from a shell) fundamentally cannot accept the "signal" (the window message sent by `.CloseMainWindow()`.

* (b) **applications that _do_ have a message loop and accept the "signal" will predictably _not_ terminate if they're GUI applications in a "dirty" state that causes them to pop up a modal confirmation dialog.**

In the case of (a), you deserve to know that _graceful_ termination is _fundamentally impossible_ - this is what `taskkill` already does.

In the case of (b), the only assurance you have is that the signal was "sent" - and there's a definite chance that termination will _not_ occur.

If you implement a fallback to killing just for (a), you introduce an awkward asymmetry that additionally _hides_ a system capability: to _request_ termination _without enforcing it_.

Conversely, you have _not_ ensured overall that termination will take place.

---

**The only unified _platform-neutral_, desired-state behavior that is worth providing is the one that would (a) require opt-in via `Force` (indicating that you want to _fall back to killing_) and (b) therefore requires a more complex implementation with timeout-based waiting**, which is the only way to _guarantee_ termination.

**Comment by iSazonov** [comment]
You seem to be ignoring the fact that the "grace stopping" is only a function of the application and it works only as the developer implemented it. There can be an infinite number of implementations. There's not even precise feedback (from the API) even on Windows. Moreover, there is not even a predefined semantics, even on Windows - the developer can assign any action to the event. There is no point in looking for something in common and trying to make a universal solution.
We can find many interactive programs that behave differently. So MS Word could ask to save a file, but VS Code couldn't. If we pay attention to services, they behave very differently from interactive applications. For example, if we send a SIGTERM to a web server, it could not mean stopping it, but a cold restart.
All we can do is send a signal and forget. We do this in Stop-Process only because we hope that, in general, application developers follow common practice for this event, however each implementation will be different for each application. We can do more only for a specific application by implementing the necessary logic in the commands following the Stop-Process in a script.

**Comment by mklement0** [comment]
> All we can do is send a signal and forget.

That's what we can and should do in the simplest case (`-TryGraceful` alone).

If sending the signal _fails_, we should report that as an error, which on Windows - unfortunately, but that can't be helped - categorically includes console applications not directly running in a console (as is typical, given that they're usually launched from a *shell*, which is the one running directly in the console).

(If sending _succeeds_, we're done and move on, as you suggest,)

As stated, we could do this in a first step, and implement ensured-termination logic (see below) later, if ever.

> We can do more only for a specific application by implementing the necessary logic in the commands following the Stop-Process in a script.

No, as proposed, there _is_ something we can do: 

_If requested_, we can _ensure_ that the process terminates by _waiting_ for termination - whether with a default or specified timeout period - and if termination hasn't occurred within that timeout - for whatever reason (we can't know) - _kill_ the process then.

Of course, anyone can implement that logic themselves using the existing capabilities, but that's fairly cumbersome, so 
we would be offering this as built-in desired-state logic, as a  courtesy.

**Comment by mklement0** [comment]
It sounds like the only sticking point is what to do if `.CloseMainWindow()` returns `$false` with `-TryGraceful` (alone):

* Approach A: Report an error and do nothing else; this is what `taskkill.exe` does.
* Approach B: Automatically and quietly fall back to killing; this is what you're advocating, if I understand correctly.

I find B problematic, but I can also see the appeal of its pragmatism.

If everyone is comfortable with it, so be it.

**Comment by iSazonov** [comment]
> It sounds like the only sticking point is what to do if .CloseMainWindow() returns $false with -TryGraceful (alone):

Yes. It is only about Windows.

> Of course, anyone can implement that logic themselves using the existing capabilities (assuming -TryGraceful alone has been implemented), but that's fairly cumbersome, so
we would be offering this as built-in desired-state logic, as a courtesy.

And yes, and no. We could do something smart if CloseMainWindow() did something smart. But see the method implementation:
https://github.com/dotnet/runtime/blob/8a52f1e948b6f22f418817ec1068f07b8dae2aa5/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Win32.cs#L240-L260

This method does nothing smart. It fast return false if no main window is or it is disabled - why do we need to write an error if such application makes no distinction between killing and grace stopping? For such application killing is the same as grace stopping.
And as we can see if CloseMainWindow() sends WM_Close event it returns always true - not a result from the application as we could expect! It is exactly as on Unix for SIGTERM. 
There is recommendation for processing WM_Close in [docs](https://docs.microsoft.com/en-us/windows/win32/learnwin32/closing-the-window) but this does not mean that everyone will do exactly that. For example, VS Code does not ask to save the file on close, unlike MS Word.

So on all OS-s the result is unpredictable - a script write is forced to _always check_ the result he expects in his particular scenario - no errors help.

**Comment by mklement0** [comment]
> It fast return false if no main window is or it is disabled - why do we need to write an error if such application makes no distinction between killing and grace stopping?

Because you may choose _not to kill_ if graceful termination isn't possible.

Consider this scenario: Notepad is open with an unsaved file, and a modal dialog happens to be displayed in it: even though graceful termination is possible with a GUI application such as Notepad _in principle_, in this case `.CloseMainWindow()` returns  `$false` and will result in _data loss_ if you fall back to killing.

With your proposal you'll get _reliable_ termination only when `.CloseMainWindow()` returns `$false`.

If it returns `$true`, given how GUI applications on Windows respond with unsaved data, there's a good chance you won't get termination at all - which you can - cumbersomely - test for manually, or, as a desired-state _convenience_ feature, we can offer this functionality via `Stop-Process` on an _opt-in_ basis; see next point.

By contrast, what I proposed would amount to `-TryGraceful` simply being a wrapper for `.CloseMainWindow()` on Windows, and its situational failure - indicated by a return value  of `$false` - would be surfaced in a PowerShell-appropriate manner, as a non-terminating error.

It is then up to the user to decide _whether killing is appropriate_. 
In the simplest case the fallback-to-kill logic _as an explicit choice_ could then be achieved with:
`Stop-Process -TryGraceful $somePid 2>$null || Stop-Process $somePid`

> We could do something smart

We can do something _predictable_ that _ensures_ a desired outcome, namely reliable termination - covering _both_ the cannot-send-signal and signal-sent-but-process-didn't-terminate scenarios, as previously described.

It would be a  _convenience_ feature that compensates for the lack of predictability of the underlying system features.

As stated,  it requires additional parameters and a more complex implementation.

**Comment by mklement0** [comment]
Let me try to bring closure to this by summarizing our options:

* Existing behavior remains as-is: target processes are _killed_ by default.

* A conceptually simple and therefore easy-to-implement solution is to implement (only) a `-TryGraceful` switch:

  * It sends a _request_ to terminate to a target process, which it may or may not honor, but that's not `Stop-Process`' concern; the request is sent:
    * in the form of the `SIGTERM` signal on Unix
    * by calling `.CloseMainWindow()` on Windows.

  *  On Unix, failing to send `SIGTERM` represents a condition that leaves reporting an error as the only option (it's likely a permissions problem, which means that `SIGKILL` cannot be sent either).

  *  On Windows, we have a _choice_ as to how to respond to `.CloseMainWindow()` indicating failure by way of returning `$false`:

      * Context:
        * `$false` is returned in two distinct scenarios:
           * the target process has no message loop, either because it isn't a GUI-subsystem application or because it is a console-subsystem application that doesn't _directly_ run in a console window (which - typically, but not necessarily - only _shell_ processes do)
           * the target process _does_ have a message loop, but is unable to receive messages at the moment, either because it hangs or because its GUI is showing a _modal dialog_ that blocks the main message loop.
  
        * `$true` just means that the message was successfully _sent_, but with GUI applications there is a distinct chance that they will _not_ terminate, namely if they're in a "dirty" state that causes them to pop up a modal dialog to ask the user to confirm the intent to terminate either with or without saving.

      * Implementation options for the case when `$false` is returned:

          * **Option 1**: Report a _non-terminating error_ and take no further action.
            * This follows the default behavior of `taskkill.exe` (using option `/f` requests _kill_ behavior).
            * If the user wants to ensure termination, they'll have to call `Stop-Process` again, _without_ `-TryGraceful`.

          * **Option 2**: Fall back to _killing_ the process, which has the following implications:
            * Most console applications (except those running directly in console windows, which are typically only shells) are _always killed_.
            * GUI applications that hang and those that happen to be showing a modal dialog are _killed_.
            * All other applications either terminate gracefully, refuse to terminate, or, in the case of GUI applications, delegate the termination decision to the user by popping up a modal confirmation dialog.

  * On _both_ platforms, if ultimate _termination_ is to be _ensured_ in _all_ cases, the user will have to use _additional commands_ in order to _wait for a while_  (`Wait-Process`) and, if waiting times out, call `Stop-Process` _without_ `-TryGraceful` to effect termination by _killing_.

* A more complex enhancement that provides _ensured-termination_ logic can be added later, which then needs to incorporate the timeout-based waiting and kill fallback into `Stop-Process` itself.

**Comment by iSazonov** [comment]
To get a consistency on all platforms we could not use `CloseMainWindow()` but our custom method without first two checks - only send WM_Close event - we would get a behavior like with SIGTERM on Unix - send and forget. (If an user want exactly `CloseMainWindow()` behavior he can call it directly.)

**Comment by mklement0** [comment]
If you really wanted to do that, you could simply ignore `.CloseMainWindow()`'s return value.

However, to me that's not consistency - that's just hiding an error condition from the user, given that he intent of _requesting_ termination could not be fulfilled.

**Comment by danielniccoli** [comment]
It is unfortunate that SIGTERM is not the default. While I understand that some people here have concerns regarding breaking changes, I think it is not wise to have expected behaviour hidden behind a switch. Breaking changes happen all the time. From 7.0 to 7.1 to 7.2 to 7.3 (preview).

I suggest to rather have a breaking change now, then forgetting to change the default behaviour in 8.0 and keep postponing this needed change for years.

**Comment by davesbrown** [comment]
Not sure if this is right thread, but I've noticed a behavior where I would wait for the ProcessId to clear (zero or null?) but recently, seems something has changed.  When I use wmiobject or ciminstance to get the PID after a stop, the PID is immediately zero, but if I run handle.exe I'll still see file locks with an associated PID.

**Comment by manokara** [comment]
I'm on Windows with PowerShell 7.1.5. This behavior is quite unintuitive, because there is a `-Force` switch but `Stop-Process` always acts as if it was on, killing a process without mercy instead of gracefully shutting it down. I've noticed this also happens with Task Manager - it doesn't give you an option to *request* a process to terminate, allowing for cleanup routines to be run. Being mostly a Linux user, I'm used to `SIGTERM`ing processes and only `SIGKILL`ing them as a last resort. Seems like `SIGKILL` is the default expected behavior on Windows.

In my use case, I'm dealing with console processes on a CI deployment pipeline so `CloseMainWindow()` doesn't apply. Using `taskkill` instead as OP mentioned works as expected.

**Comment by kapsiR** [comment]
Dear @microsoft-github-policy-service, please don't close this issue - it's still a thing

**Comment by alx9r** [comment]
I just read through most of this remarkably informative discussion.  For what it's worth, here is my intuition for overcoming these rough edges in idiomatic PowerShell:

- `Stop-Process` - I would leave this behavior more-or-less as is.  "Stop" has a rather clear imperative implication and doesn't imply (to me, at least) that the command is negotiable by the process.  In other words, I expect `Stop-Process` to stop the process or report an error.  Accordingly, I think the current behavior (at least on Windows) is more-or-less as expected.  I naturally expect `-Force` to make `Stop-Process` try something more drastic where it otherwise fails to stop the process with an error.  That is how `Expand-Archive` works: File collisions produce an error, and `-Force` clobbers if you want that.  
- `Request-Process Stop` - I propose this command be added that does all the requesting that `Stop-Process` does but none of the forceful, possibly-destructive, killing that `Stop-Process` can do.  I suggest that this command could implement some of the numerous somewhat-common methods of signaling a request to stop that any given process on Windows may or may not respond to.

**Comment by danielniccoli** [comment]
> - `Request-Process Stop` - I propose this command be added that does all the requesting that `Stop-Process` does but none of the forceful, possibly-destructive, killing that `Stop-Process` can do.

Request: Asks for a resource or asks for permissions.

Stop: Discontinues an activity. This verb is paired with Start.

https://learn.microsoft.com/de-de/powershell/scripting/developer/cmdlet/approved-verbs-for-windows-powershell-commands?view=powershell-7.4#lifecycle-verbs

Stop is the correct verb and Request is meant for unidirectional information flow. Using it to send a command back to the process would be  unintuitive.

**Comment by alx9r** [comment]
>Request: Asks for a resource or asks for permissions.

@danielniccoli I noticed that too.  I think the counterarguments would be the following:

1. Few people read those definitions.
2. It's not obvious that those definition are exclusive.
3. In customary english the meaning of "request" is far broader than that definition.

>Using it to send a command back to the process would be unintuitive.

It's not unintuitive to me.

I would note that that list of approved verbs is just another example of PowerShell doctrine.  Doctrine is valuable when it's reliably right.  Unfortunately there are endless examples where PowerShell doctrine has been unhelpful, misguided, counterproductive, and downright incorrect.  As a result, I generally find arguments based on PowerShell doctrine alone entirely unpersuasive.

**Comment by danielniccoli** [comment]
> 1. Few people read those definitions.

This is a weak argument. Even fewer people read RFCs, yet the Internet only functions because protocol implementations and applications adhere to them.

> 2. It's not obvious that those definition are exclusive.

The developers of all built-in cmdlets seem to follow them to the point, making this argument a you-problem.

> 3. In customary english the meaning of "request" is far broader than that definition.

See above. The documentation makes it pretty clear how it should be used.

**Comment by alx9r** [comment]
>It doesn't add anything productive to the discussion about this particular topic.

I think this last sentence is a fair criticism.  Let me elaborate, then.  The problem with strictly sticking to the description of the approved verbs has been that, because most are each such narrow descriptions, commands that did something slightly outside the description became `Invoke-` because it was the only verb whose description actually fit.  The command I proposed requests but does not force a stop.  The description for `Get-Verb Stop` is "Discontinues an activity" which, strictly speaking, a command that requests but does not force a process to stop does not fit.  So, by the same argument that `Request-Process` uses a verb where the command's behavior does not fit the description, `Stop-Process` also does not.  Which leaves us with something like `Invoke-ProcessRequest`.  At least `Invoke-ProcessRequest` would be consistent with `Invoke-WebRequest`.  But I don't think even more instances of the meaning-free `Invoke-` verb helps readability.

**Comment by danielniccoli** [comment]
> The description for Get-Verb Stop is "Discontinues an activity" which, strictly speaking, a command that requests but does not force a process to stop does not fit.

I disagree with your assertion that Stop-Process doesn’t align with the verb definition of “discontinues an activity.” Stop conveys intent clearly: the cmdlet's purpose is to discontinue the process. Whether the action is graceful or forceful is determined by parameters like -Force. This is entirely consistent with PowerShell conventions.

It’s also important to note that the list of approved verbs does not guarantee the outcome of a cmdlet—it only conveys the intent of the action. For example, Stop-Process signals the intent to terminate a process, but the outcome depends on factors like permissions, process behavior, or the use of parameters like -Force. This focus on intent is why Stop is a better fit for a cmdlet like this than Request.

**Comment by alx9r** [comment]
I think we can disagree in good faith about this.  I'm confident we can each always find counterexamples to whatever assertion is made about how existing PowerShell works or should work.  This is just another example of where the guidelines has led to valuable time spent arguing over "How many angels can dance on the head of a pin?"  I certainly prefer consistent API naming, but for me, personally, the cost in time debating naming given those guidelines is as Sisyphean as trying to make any broad statements about [PowerShell's error behavior](https://github.com/MicrosoftDocs/PowerShell-Docs/issues/1583).  I'm very interested in functionality, but these last few posts have reinforced that I prefer to be left out of command naming discussions.

**Comment by danielniccoli** [comment]
@alx9r, while I appreciate your confidence in disagreeing in good faith, your response sidesteps the actual issue. This isn’t a philosophical debate about “angels on the head of a pin”; it’s about the practical importance of following established guidelines to ensure PowerShell cmdlets are consistent, clear, and user-friendly.

Dismissing this discussion as "Sisyphean" undermines the value of naming conventions. Consistent naming is not just theoretical—it directly impacts usability. Poorly chosen or inconsistent names create confusion for users and increase the difficulty of learning and working with PowerShell. You’ve acknowledged that you prefer consistent API naming, but opting out of this discussion while advocating for a deviation contradicts that stance.

The guidelines exist to minimize these debates and ensure discoverability for all users, not just those who design commands. Ignoring them isn’t a matter of personal preference—it has consequences for the wider community. While you may prefer to focus solely on functionality, it’s worth remembering that naming is functionality when it affects how users interact with the tool.

**Comment by alx9r** [comment]
@danielniccoli  I disagree with many of the sentiments in your last comment, but this isn't the place for that, and I'm not interested in the topic until the conditions for resolving such disagreements is improved.  That's not a bridge I'm going to build, but I'll appreciate and cross it if and when it's built.

**Comment by alx9r** [comment]
Further to [this comment](https://github.com/PowerShell/PowerShell/issues/13664#issuecomment-774185156)

>It seems that Windows currently fundamentally doesn't support sending a terminating request...

Indeed _Windows_ seems to provide no such standardized facility and, accordingly, Windows applications seem to not listen for such a request.  The console of a PowerShell process, on the other hand, does seem to robustly respond to `CTRL_C_EVENT` raised from within its process.  `CTRL_C_EVENT` seems to be get mapped to the [`PosixSignal` events](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.posixsignal?view=net-8.0).  (Whether that's `SIGINT`, `SIGTERM`, otherwise is, I think, moot for the PowerShell console on Windows.)  One PowerShell process can connect to another PowerShell process and run things in an in-process runspace.  This all means that one PowerShell process can reliably send console control events to another PowerShell process.  

How to do PowerShell to PowerShell `SIGINT`, `SIGTERM` or `CTRL_C_EVENT` on Windows was not at all obvious to me on my first read through this thread.  I know that's not exactly what the OP is asking, but many of my searches for that capability led here, so I'll leave this example here, in case it helps anyone.

Also related is [this technique](https://stackoverflow.com/a/79249620/1404637) which stops all the runspaces in the target process, instead.  That technique works on jobs where the console control events do not.

### Demo Code
<details>
<summary>
This code creates a second `pwsh` process with a console that's blocked in `Start-Sleep`.  It then connects to that second `pwsh` process and raises `CTRL_C_EVENT` which causes the sleep to be cancelled and the process to gracefully end.
</summary>

```PowerShell
function Send-PowerShellConsoleControlEvent {
    param(
        [Parameter(Mandatory,ValueFromPipelineByPropertyName)]
        [int]
        [Alias('Id')]
        $ProcessId,

        [Parameter(Mandatory)]
        [int]
        $CtrlEvent,

        [Parameter()]
        [int]
        $ProcessGroupId = 0
    )
    process {
        $pipe = [System.Management.Automation.Runspaces.NamedPipeConnectionInfo]::new($ProcessId)
        $rs = [runspacefactory]::CreateRunspace($pipe)
        $rs.Open()
        $ps =
            [powershell]::Create($rs).
            AddScript({
                param($CtrlEvent,$ProcessGroupId)
                try {
                    Add-Type @'
namespace n { public static class c {
        [System.Runtime.InteropServices.DllImport("KERNEL32.dll",ExactSpelling=true,SetLastError=true)]
        public static extern bool GenerateConsoleCtrlEvent(int dwCtrlEvent,int dwProcessGroupId);
    }
}
'@
                    [n.c]::GenerateConsoleCtrlEvent($dwCtrlEvent,$ProcessGroupId)
                }
                catch {
                    $_
                }
            }).
            AddArgument($CtrlEvent).
            AddArgument($ProcessGroupId)
        $ps.Invoke()
    }
    clean {
        ${rs}?.Dispose()
        ${ps}?.Dispose()
    }
}

$process =
    Start-Process                   `
        -PassThru                   `
        -FilePath     'pwsh'        `
        -ArgumentList '-NoProfile'  ,
                      '-Command'    ,
                      { Start-Sleep 100}
Start-Sleep 1
$process | Send-PowerShellConsoleControlEvent -CtrlEvent 0

$process | Wait-Process -Timeout 1
```

</details>
